xarray/core/dataarray.py

```diff
@@ -1374,5 +1374,64 @@ def real(self):
     def imag(self):
         return self._replace(self.variable.imag)
 
+    def dot(self, other):
+        """Perform dot product of two DataArrays along their shared dims.
+        
+        Equivalent to taking taking tensordot over all shared dims.
+
+        Parameters
+        ----------
+        other : DataArray
+            The other array with which the dot product is performed.
+            
+        Returns
+        -------
+        result : DataArray
+            Array resulting from the dot product over all shared dimensions.
+            
+        See also
+        --------
+        np.tensordot(a, b, axes)
+
+        Examples
+        --------
+
+        >>> da_vals = np.arange(6 * 5 * 4).reshape((6, 5, 4))
+        >>> da = DataArray(da_vals, dims=['x', 'y', 'z'])    
+        >>> dm_vals = np.arange(4)
+        >>> dm = DataArray(dm_vals, dims=['z'])
+                
+        >>> dm.dims
+        ('z')
+        >>> da.dims
+        ('x', 'y', 'z')
+
+        >>> dot_result = da.dot(dm)
+        >>> dot_result.dims
+        ('x', 'y')
+        """
+        if isinstance(other, Dataset):
+            raise NotImplementedError('dot products are not yet supported '
+                                      'with Dataset objects.')
+        if not isinstance(other, DataArray):
+            raise TypeError('dot only operates on DataArrays.')
```

Comment:
Yeah, could go either way here. I like raising `NotImplementedError` because it makes it more obvious to users that this behavior might change in the future. Besides, if we get lucky a user will notice the error and then implement the missing functionality themselves ;).